internal/services/directoryroles/directory_role_templates_data_source.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package directoryroletemplates
+
+import (
+	"context"
+	"crypto/sha1"
+	"encoding/base64"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-provider-azuread/internal/clients"
+	"github.com/hashicorp/terraform-provider-azuread/internal/tf"
+)
+
+func directoryRoleTemplatesDataSource() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: directoryRoleTemplatesDataSourceRead,
+
+		Timeouts: &schema.ResourceTimeout{
+			Read: schema.DefaultTimeout(5 * time.Minute),
+		},
+
+		Schema: map[string]*schema.Schema{
+			"object_ids": {
+				Description: "The object IDs of the role templates",
+				Type:        schema.TypeList,
+				Computed:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+
+			"role_templates": {
+				Description: "A list of role templates",
+				Type:        schema.TypeList,
+				Computed:    true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"display_name": {
+							Description: "The display name of the directory role template",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+
+						"description": {
+							Description: "The description of the directory role template",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+
+						"object_id": {
+							Description: "The object ID of the directory role template",
+							Type:        schema.TypeString,
+							Computed:    true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func directoryRoleTemplatesDataSourceRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*clients.Client).DirectoryRoleTemplates.DirectoryRoleTemplatesClient
+
+	directoryRoleTemplates, _, err := client.List(ctx)
+	if err != nil {
+		return tf.ErrorDiagF(err, "Could not retrieve role templates")
+	}
+	if directoryRoleTemplates == nil {
+		return tf.ErrorDiagF(errors.New("API error: nil directoryRoleTemplates were returned"), "Retrieving all directory role templates")
+	}
+
+	objectIds := make([]string, 0)
+	templateList := make([]map[string]interface{}, 0)
+
+	for _, r := range *directoryRoleTemplates {
```

Comment:
Should we be filtering out the `Users` role template? Per the docs:

> Note: A directory role template is exposed for the Users directory role. The Users directory role is implicit and is not visible to directory clients. Every User in the tenant is assigned to this role by the infrastructure. The role is already activated. Do not use this template.

Reply:
Excluded "Users"